src/lazyfree.c

```diff
@@ -143,14 +143,16 @@ size_t lazyfreeGetFreeEffort(robj *key, robj *obj, int dbid) {
  * will be reclaimed in a different bio.c thread. */
 #define LAZYFREE_THRESHOLD 64
 int dbAsyncDelete(redisDb *db, robj *key) {
+    dictEntry **plink;
+    int table;
     /* Deleting an entry from the expires dict will not free the sds of
      * the key, because it is shared with the main dictionary. */
     if (dictSize(db->expires) > 0) dictDelete(db->expires,key->ptr);
 
     /* If the value is composed of a few allocations, to free in a lazy way
      * is actually just slower... So under a certain limit we just free
      * the object synchronously. */
-    dictEntry *de = dictUnlink(db->dict,key->ptr);
+    dictEntry *de = dictFindWithPlink(db->dict,key->ptr,&plink, &table);
```

Comment:
we wanna move the call for `moduleNotifyKeyUnlink` to be before the dictDelete from the expires dict.
one way to do that is  to move both `moduleNotifyKeyUnlink` and `dictFindWithPlink` to the top of the function, but maybe it's easier and actually better to move the deletion from the expires dict into the `if (de)` block below.

Reply:
There are two `if`, I think we should merge them